lib/auth/userloginstate/generator.go

```diff
@@ -0,0 +1,188 @@
+	if err := config.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &Generator{
+		accessLists: config.AccessLists,
+		access:      config.Access,
+		clock:       config.Clock,
+	}, nil
+}
+
+// Generate will generate the user login state for the given user.
+func (g *Generator) Generate(ctx context.Context, user types.User) (*userloginstate.UserLoginState, error) {
+	// Create a new empty user login state.
+	uls, err := userloginstate.New(
+		header.Metadata{
+			Name: user.GetName(),
+		}, userloginstate.Spec{
+			Roles:  []string{},
+			Traits: trait.Traits{},
+		})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Generate the user login state.
+	if err := g.addAccessListsToState(ctx, user, uls); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Clean up the user login state after generating it.
+	if err := g.postProcess(ctx, uls); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return uls, nil
+}
+
+// addAccessListsToState will added the user's applicable access lists to the user login state.
+func (g *Generator) addAccessListsToState(ctx context.Context, user types.User, state *userloginstate.UserLoginState) error {
+	accessLists, err := g.accessLists.GetAccessLists(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Create an identity for testing membership to access lists.
+	identity := tlsca.Identity{
+		Username: user.GetName(),
+		Groups:   user.GetRoles(),
+		Traits:   user.GetTraits(),
+	}
+
+	for _, accessList := range accessLists {
+		if err := services.IsMember(identity, g.clock, accessList); err != nil {
+			continue
+		}
+
+		state.Spec.Roles = append(state.Spec.Roles, accessList.Spec.Grants.Roles...)
+
+		for k, values := range accessList.Spec.Grants.Traits {
+			state.Spec.Traits[k] = append(state.Spec.Traits[k], values...)
+		}
+	}
+
+	return nil
+}
+
+// postProcess will perform cleanup to the user login state after its generation.
+func (g *Generator) postProcess(ctx context.Context, state *userloginstate.UserLoginState) error {
+	// Deduplicate roles and traits
+	state.Spec.Roles = utils.Deduplicate(state.Spec.Roles)
+
+	for k, v := range state.Spec.Traits {
+		state.Spec.Traits[k] = utils.Deduplicate(v)
+	}
+
+	// If there are no roles, don't bother filtering out non-existent roles
+	if len(state.Spec.Roles) == 0 {
+		return nil
+	}
+
+	// Remove roles that don't exist in the backend so that we don't generate certs for non-existent roles.
+	// Doing so can prevent login from working properly. This could occur if access lists refer to roles that
+	// no longer exist, for example.
+	roles, err := g.access.GetRoles(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	roleLookup := map[string]bool{}
+	for _, role := range roles {
+		roleLookup[role.GetName()] = true
+	}
+
+	existingRoles := []string{}
+	for _, role := range state.Spec.Roles {
+		if roleLookup[role] {
+			existingRoles = append(existingRoles, role)
+		}
+	}
+	state.Spec.Roles = existingRoles
```

Comment:
I think we should warn that given roles were granted but do not exist. This may be due to the role being removed, but may well be due to a typo in the definition. Having a warning here is a step toward finding out about that typo.